File: buzu/src/main/java/br/net/buzu/util/StaticBehave.java
/*
 *	This file is part of Buzu.
 *
 *   Buzu is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   Buzu is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with Buzu.  If not, see <http://www.gnu.org/licenses/>.
 */
package br.net.buzu.util;

import br.net.buzu.model.MetaInfo;
import br.net.buzu.model.Metadata;
import br.net.buzu.model.StaticStructure;

import java.util.Collection;

/**
 * Util methods for Static behave.
 *
 * @author Douglas Siviotti
 * @since 1.0
 *
 */
public final class StaticBehave {

	static final String META_INFO_MUST_BE_COMPLETE = "MetaInfo must be complete (size and occurrences) for static behave:";
	static final String META_INFO_CANNOT_BE_UNBOUNDED = "MetaInfo cannot be Unbounded (no limit) for static behave:";
	static final String INVALID_STATIC_CHILD = "Invalid Static child:";

	private StaticBehave() {
	}

	public static MetaInfo checkStaticInfo(MetaInfo metaInfo) {
		if (!metaInfo.isComplete()) {
			throw new IllegalArgumentException(META_INFO_MUST_BE_COMPLETE + metaInfo);
		}
		if (metaInfo.isUnbounded()) {
			throw new IllegalArgumentException(META_INFO_CANNOT_BE_UNBOUNDED + metaInfo);
		}
		return metaInfo;
	}

	public static void checkStaticChild(Metadata child) {
		if (!(child instanceof StaticStructure)) {
			throw new IllegalArgumentException(
					INVALID_STATIC_CHILD + child.info()+ " is not a " + StaticStructure.class.getSimpleName());
		}
	}

	public static boolean isStaticChidren(Collection<? extends Metadata> children){
		for (Metadata child : children){
			if (!child.isStatic()){
				return false;
			}
		}
		return true;
	}

	public static boolean isStatic(Metadata metadata){
		return metadata.hasChildren()? isStaticChidren(metadata.children()): metadata.info().isStatic();
	}


}

File: ppl-lib/src/main/kotlin/br/net/buzu/lib/StaticBehave.kt
/*
 *	This file is part of Buzu.
 *
 *   Buzu is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   Buzu is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with Buzu.  If not, see <http://www.gnu.org/licenses/>.
 */
package br.net.buzu.lib

import br.net.buzu.java.model.MetaInfo
import br.net.buzu.java.model.Metadata
import br.net.buzu.java.model.StaticStructure

internal val META_INFO_MUST_BE_COMPLETE = "MetaInfo must be complete (size and occurrences) for static behave:"
internal val META_INFO_CANNOT_BE_UNBOUNDED = "MetaInfo cannot be Unbounded (no limit) for static behave:"
internal val INVALID_STATIC_CHILD = "Invalid Static child:"


fun checkStaticInfo(metaInfo: MetaInfo): MetaInfo {
    if (!metaInfo.isComplete) {
        throw IllegalArgumentException(META_INFO_MUST_BE_COMPLETE + metaInfo)
    }
    if (metaInfo.isUnbounded) {
        throw IllegalArgumentException(META_INFO_CANNOT_BE_UNBOUNDED + metaInfo)
    }
    return metaInfo
}


fun checkStaticChild(child: Metadata) {
    if (child !is StaticStructure) {
        throw IllegalArgumentException(
                INVALID_STATIC_CHILD + child.info() + " is not a " + StaticStructure::class.java.simpleName)
    }
}

fun isStaticChildren(children: List<Metadata>): Boolean {
    for (child in children) {
        if (!child.isStatic()) {
            return false
        }
    }
    return true
}

fun complexSize(children: List<Metadata>): Int {
    var complexSize = 0
    for (child in children) {
        complexSize += child.info().size
    }
    return complexSize
}